### ISSUE TITLE
Miscalculated diff

### ISSUE THREAD

**Issue by mattiasdrp** [janestreet/patdiff#19]
With the two following files:

`notmp`

    x: int
    y: int

`tmp`

    x: int
    y: int
    tmp: int

patdiff is able to compute the proper diff:

![image](https://user-images.githubusercontent.com/5543639/166714466-112402fe-de31-463a-919d-bf1152175348.png)

But if I put punctuation in it:

`notmp`

    x: int
    y: int

`tmp`

    x: int;
    y: int;
    tmp: int

The calculated diff doesn't seem to be the optimal one:

![image](https://user-images.githubusercontent.com/5543639/166714620-02ca49b1-deda-430a-bb93-4168f3d4ebe7.png)

Is this the expected behaviour?

Since this is my third issue, thanks for the package btw, sorry for the spam!

**Comment by bcc32** [comment]
Yeah, this is a known behavior, if not the ideal one.  I don't have any terribly good ideas for fixing it, unfortunately, since the way patdiff currently does word-level diffs involves combining adjacent lines and "discarding" the newlines in order to calculate the diff, and it just so happens to pick a diff region that crosses more line boundaries.

**Comment by mattiasdrp** [comment]
Isnt't there a way to give a higher weight to a diff containing a newline than one containing none? I don't know how the algorithm works so I'm just speculating.

**Comment by bcc32** [comment]
Patdiff doesn't really use the usual "longest common subsequence" diff algorithm you might be thinking of (except as a fallback in certain circumstances).  The word-level diffs code I'm talking about is the `Patdiff_core.refine` function, whose brief internal description is:

```
  (* Refines the diff, splitting the lines into smaller arrays and diffing them, then
     collapsing them back into their initial lines after applying a format. *)
```

Feel free to browse the `refine` function to get an idea of what it does.  I don't think there's an obvious place to inject the heuristic you suggested, but it's quite possible there's something I've missed.

**Comment by mattiasdrp** [comment]
I'll take a look if I find the time for it but thanks for the comment and the second look :-)
